### ISSUE TITLE
RAM requirement

### ISSUE THREAD

**Issue by RichardCorbett** [nanoporetech/bonito#97]
Hi there,

I'm new to using GPUs so this could be a very basic question.    I am finding that I am running out of RAM with my bonito jobs.   
My command:
`CUDA_VISIBLE_DEVICE=0; /opt/bonito-0.3.2/bin/bonito basecaller --fastq dna_r9.4.1 04_split > bonito_0_3_2_PAG44265_04_split/PAG44265_04_split.fastq`

My slurm job requests:
```
#SBATCH --mem 65536M
#SBATCH -n 1 --gres=gpu:1 --partition=redacted
```

but I get error messages like this:
`error: Job 107820 exceeded memory limit (69823960 > 67108864), being killed`
suggesting I need to give more than 64Gb of RAM to each job I submit.   Does this make sense?   In case it helps, I can see that my jobs do seem to be getting on to the GPUs (via gpustat).   What is the recommended amount of non-GPU RAM recommended to run bonito?  The GPU RAM requirement seem to be around 5Gb per job.

**Comment by RichardCorbett** [comment]
Still pushing up the RAM limit to 124Gb and getting the same error.

**Comment by iiSeymour** [comment]
Hi @RichardCorbett

We had a memory leak which I think should be fixed on master - are you able to test that?

**Comment by RichardCorbett** [comment]
For sure.  Will set it up now.

**Comment by RichardCorbett** [comment]
Do you have a ballpark amount of RAM that would be a conservative request for me to use for the test jobs?

**Comment by RichardCorbett** [comment]
Using your current master we ran the command above and got this output until we hit 32Gb of RAM and the cluster killed the job.
`  > calling: 0 reads [00:00, ? reads/s]
  > calling: 1 reads [00:31, 31.86s/ reads]
  > calling: 7 reads [00:36, 22.55s/ reads]
  > calling: 9 reads [00:37, 15.86s/ reads]
  > calling: 10 reads [00:42, 12.77s/ reads]
  > calling: 16 reads [00:47,  9.16s/ reads]
  > calling: 28 reads [00:54,  6.60s/ reads]
  > calling: 32 reads [01:01,  5.10s/ reads]
  > calling: 39 reads [01:04,  3.70s/ reads]
  > calling: 58 reads [01:04,  2.59s/ reads]
  > calling: 64 reads [01:09,  2.07s/ reads]
  > calling: 68 reads [01:12,  1.67s/ reads]`

**Comment by menickname** [comment]
Dear @iiSeymour 

I am having the same issue when running bonito (v.0.3.2) on our HPC. I gave bonito 256Gb or RAM, however this was not enough and the run got killed after about 7000 reads were basecalled. 

@boegel supplied me with the following HPC log files:

`2020-12-21T12:43:22.131191+01:00 node3302 kernel: Cannot map memory with base addr 0x2ba4f4000000 and size of 0x6031 pages
2020-12-21T12:43:22.131238+01:00 node3302 kernel: bonito invoked oom-killer: gfp_mask=0xd0, order=0, oom_score_adj=0
2020-12-21T12:43:22.131542+01:00 node3302 kernel: memory: usage 268435456kB, limit 268435456kB, failcnt 53477`

Thank you in advance.

**Comment by menickname** [comment]
Dear @iiSeymour 

We have installed the latest master (commit ID f1e46d9) and still got the excessive memory issue. Any further update on this? I did run two separate jobs on HPC and got following errors:

Run 1:
`slurmstepd: error: Detected 1 oom-kill event(s) in step 40102424.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.`

Run2:
`RuntimeError: cuda runtime error (1) : invalid argument at ../aten/src/THC/THCCachingHostAllocator.cpp:278`

Both jobs got killed after about 45 minutes which corresponds to only 14,000-17,000 reads being basecalled.

Thank you in advance.

**Comment by iiSeymour** [comment]
Thanks both for confirming the issue is still seen with the latest commit - it's a bit of tricky one, I will provide an update when I have more details.

**Comment by RichardCorbett** [comment]
Noticed that guppy 4.4.* now handles bonito-like models.      I am able to run that with reasonable RAM resources and have so far found the results to be comparable to my shallow bonito results, so might make this ticket obsolete.  @iiSeymour, is using guppy with bonito models the recommended route to take advantage of high quality basecalls moving forward?

**Comment by iiSeymour** [comment]
Yes, the same model weights from Bonito are converted from PyTorch to the format that Guppy uses so the results should be very similar.

**Comment by menickname** [comment]
dear @iiSeymour , this means the issue is solved? Or do you recommend to go for Guppy as indicated by @RichardCorbett? I was waiting for the update as you told me on December 23th. A dataset is waiting to be analysed as soon as possible!

Thanks!

**Comment by iiSeymour** [comment]
Hey @menickname - yes the issue was closed by commit 7796927ef60eebf026851c57f8feb35301e44378 as the problem is now solved, please give it try.
